pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/mutablefst/MutableFSTImpl.java

```diff
@@ -0,0 +1,198 @@
+  private MutableState _start;
+
+  public MutableFSTImpl() {
+    _start = new MutableState(true);
+  }
+
+  /**
+   * Get the initial states
+   */
+  @Override
+  public MutableState getStartState() {
+    return _start;
+  }
+
+  /**
+   * Set the initial state
+   *
+   * @param start the initial state
+   */
+  @Override
+  public void setStartState(MutableState start) {
+    if (_start != null) {
+      throw new IllegalStateException("Cannot override a start state");
+    }
+
+    _start = start;
+  }
+
+  public MutableState newStartState() {
+    return newStartState();
+  }
+
+  public MutableArc addArc(MutableState startState, int outputSymbol, MutableState endState) {
+    MutableArc newArc = new MutableArc(outputSymbol,
+                                        endState);
+    startState.addArc(newArc);
+    endState.addIncomingState(startState);
+    return newArc;
+  }
+
+  @Override
+  public void throwIfInvalid() {
+    Preconditions.checkNotNull(_start, "must have a start state");
+  }
+
+  @Override
+  public void addPath(String word, int outputSymbol) {
+    MutableState state = getStartState();
+
+    if (state == null) {
+      throw new IllegalStateException("Start state cannot be null");
+    }
+
+    List<MutableArc> arcs = state.getArcs();
+
+    boolean isFound = false;
+
+    for (MutableArc arc : arcs) {
+      if (arc.getNextState().getLabel() == word.charAt(0)) {
+        state = arc.getNextState();
+        isFound = true;
+        break;
+      }
+    }
+
+    int foundPos = -1;
+
+    if (isFound) {
+      Pair<MutableState, Integer> pair = findPointOfDiversion(state, word, 0);
+
+      if (pair == null) {
+        // Word already exists
+        return;
+      }
+
+      foundPos = pair.getRight();
+      state = pair.getLeft();
+    }
+
+    for (int i = foundPos + 1; i < word.length(); i++) {
+      MutableState nextState = new MutableState();
+
+      nextState.setLabel(word.charAt(i));
+
+      int currentOutputSymbol = -1;
+
+      if (i == word.length() - 1) {
+        currentOutputSymbol = outputSymbol;
+      }
+
+      MutableArc mutableArc = new MutableArc(currentOutputSymbol, nextState);
+      state.addArc(mutableArc);
+
+      state = nextState;
+    }
+
+    state.setIsTerminal(true);
+  }
+
+  private Pair<MutableState, Integer> findPointOfDiversion(MutableState mutableState,
+      String word, int currentPos) {
+    if (currentPos == word.length() - 1) {
+      return null;
+    }
+
+    if (mutableState.getLabel() != word.charAt(currentPos)) {
+      throw new IllegalStateException("Current state needs to be part of word path");
+    }
+
+    List<MutableArc> arcs = mutableState.getArcs();
+
+    for (MutableArc arc : arcs) {
+      if (arc.getNextState().getLabel() == word.charAt(currentPos + 1)) {
+        return findPointOfDiversion(arc.getNextState(), word, currentPos + 1);
+      }
+    }
+
+    return Pair.of(mutableState, currentPos);
+  }
```

Comment:
this non tail-recursion is quite dangerous because it could result in very deep stacks, what is the bound on the number of states?


I wonder if you could rework this slightly so instead of matching `List<MutableArc>` - each of which holds a `char` - and a `String` to a direct comparison between two `char[]` (convert the string to char[] at the edge, invert `List<MutableArc>`) so you can use `Arrays.mismatch`.

Reply:
Moved to an iterative model